lib/chef/provider.rb

```diff
@@ -175,6 +175,58 @@ def converge_by(descriptions, &block)
       converge_actions.add_action(descriptions, &block)
     end
 
+    #
+    # Handle patchy convergence safely.
+    #
+    # - Does *not* call the block if the current_resource's properties match
+    #   the properties the user specified on the resource.
+    # - Calls the block if current_resource does not exist
+    # - Calls the block if the user has specified any properties in the resource
+    #   whose values are *different* from current_resource.
+    # - Does *not* call the block if why-run is enabled (just prints out text).
+    # - Prints out automatic green text saying what properties have changed.
+    #
+    # @param properties An optional list of property names (symbols). If not
+    #   specified, `new_resource.class.state_properties` will be used.
+    # @param converge_block The block to do the converging in.
+    #
+    # @return [Boolean] whether the block was executed.
+    #
+    def converge_if_changed(*properties, &converge_block)
+      properties = new_resource.class.state_properties.map { |p| p.name } if properties.empty?
```

Comment:
I think resource authors will at some point type `converge_if_changed nil` and `converge_if_changed do ... end` and `converge_if_changed`.  When they write this in their resource and go test it, I think they are going to get some confusing errors raised.  A `no method to_sym on Nil` error is more confusing than an error that says `nil is not a valid property name`.

The latter clearly tells a new resource author what they need to fix to make their recipe work, while the former leaves them scratching their head.  Thats why I would suggest argument validation.